ecs-agent/daemonimages/csidriver/driver/node_linux.go

```diff
@@ -18,10 +18,151 @@ package driver
+		}
+
+		klog.V(5).InfoS("[Debug] The canonical device path was resolved", "devicePath", devicePath, "cacanonicalDevicePath", canonicalDevicePath)
+		return d.appendPartition(canonicalDevicePath, partition), nil
+	}
+
+	klog.V(5).InfoS("[Debug] Falling back to nvme volume ID lookup", "devicePath", devicePath)
+
+	// AWS recommends identifying devices by volume ID
+	// (https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/nvme-ebs-volumes.html),
+	// so find the nvme device path using volume ID. This is the magic name on
+	// which AWS presents NVME devices under /dev/disk/by-id/. For example,
+	// vol-0fab1d5e3f72a5e23 creates a symlink at
+	// /dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol0fab1d5e3f72a5e23
+	nvmeName := "nvme-Amazon_Elastic_Block_Store_" + strings.Replace(volumeID, "-", "", -1)
+
+	nvmeDevicePath, err := findNvmeVolume(d.deviceIdentifier, nvmeName)
+
+	if err == nil {
+		klog.V(5).InfoS("[Debug] successfully resolved", "nvmeName", nvmeName, "nvmeDevicePath", nvmeDevicePath)
+		canonicalDevicePath = nvmeDevicePath
+		return d.appendPartition(canonicalDevicePath, partition), nil
+	} else {
+		klog.V(5).InfoS("[Debug] error searching for nvme path", "nvmeName", nvmeName, "err", err)
+	}
+
+	if canonicalDevicePath == "" {
+		return "", errNoDevicePathFound(devicePath, volumeID)
+	}
+
+	canonicalDevicePath = d.appendPartition(canonicalDevicePath, partition)
+	return canonicalDevicePath, nil
+}
+
+func errNoDevicePathFound(devicePath, volumeID string) error {
+	return fmt.Errorf("no device path for device %q volume %q found", devicePath, volumeID)
+}
+
+// findNvmeVolume looks for the nvme volume with the specified name
+// It follows the symlink (if it exists) and returns the absolute path to the device
+func findNvmeVolume(deviceIdentifier DeviceIdentifier, findName string) (device string, err error) {
+	p := filepath.Join("/dev/disk/by-id/", findName)
+	stat, err := deviceIdentifier.Lstat(p)
+	if err != nil {
+		if os.IsNotExist(err) {
+			klog.V(5).InfoS("[Debug] nvme path not found", "path", p)
+			return "", fmt.Errorf("nvme path %q not found", p)
+		}
+		return "", fmt.Errorf("error getting stat of %q: %w", p, err)
+	}
+
+	if stat.Mode()&os.ModeSymlink != os.ModeSymlink {
+		klog.InfoS("nvme file found, but was not a symlink", "path", p)
+		return "", fmt.Errorf("nvme file %q found, but was not a symlink", p)
+	}
+	// Find the target, resolving to an absolute path
+	// For example, /dev/disk/by-id/nvme-Amazon_Elastic_Block_Store_vol0fab1d5e3f72a5e23 -> ../../nvme2n1
+	resolved, err := deviceIdentifier.EvalSymlinks(p)
+	if err != nil {
+		return "", fmt.Errorf("error reading target of symlink %q: %w", p, err)
+	}
+
+	if !strings.HasPrefix(resolved, "/dev") {
+		return "", fmt.Errorf("resolved symlink for %q was unexpected: %q", p, resolved)
+	}
+
+	return resolved, nil
+}
+
+// IsBlock checks if the given path is a block device
+func (d *nodeService) IsBlockDevice(fullPath string) (bool, error) {
```

Comment:
nit: same util method here - https://github.com/aws/amazon-ecs-agent/blob/master/ecs-agent/daemonimages/csidriver/util/utils_linux.go#L22

Reply:
(will add in a future pr with the updated build process)